test/stream_cleanup_test.go

```diff
@@ -55,3 +57,66 @@ func (s) TestStreamCleanup(t *testing.T) {
 		t.Fatalf("should succeed, err: %v", err)
 	}
 }
+
+func (s) TestStreamCleanupAfterSendStatus(t *testing.T) {
+	const initialWindowSize uint = 70 * 1024 // Must be higher than default 64K, ignored otherwise
+	const bodySize = 2 * initialWindowSize   // Something that is not going to fit in a single window
+
+	serverReturnedStatus := make(chan struct{})
+
+	ss := &stubServer{
+		fullDuplexCall: func(stream testpb.TestService_FullDuplexCallServer) error {
+			defer func() {
+				close(serverReturnedStatus)
+			}()
+			return stream.Send(&testpb.StreamingOutputCallResponse{
+				Payload: &testpb.Payload{
+					Body: make([]byte, bodySize),
+				},
+			})
+		},
+	}
+	if err := ss.Start([]grpc.ServerOption{grpc.MaxConcurrentStreams(1)}, grpc.WithInitialWindowSize(int32(initialWindowSize))); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	// This test makes sure we don't delete stream from server transport's
+	// activeStreams list too aggressively.
+
+	// 1. Make a long living stream RPC. So server's activeStream list is not
+	// empty.
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	stream, err := ss.client.FullDuplexCall(ctx)
+	if err != nil {
+		t.Fatalf("FullDuplexCall= _, %v; want _, <nil>", err)
+	}
+
+	// 2. Wait for service handler to return status.
+	//
+	// This will trigger a stream cleanup code, which will eventually remove
+	// this stream from activeStream.
+	//
+	// But the stream removal won't happen because it's supposed to be done
+	// after the status is sent by loopyWriter, and the status send is blocked
+	// by flow control.
+	<-serverReturnedStatus
+
+	// 3. GracefulStop (besides sending goaway) checks the number of
+	// activeStreams.
+	//
+	// It will close the connection if there's no active streams. This won't
+	// happen because of the pending stream. But if there's a bug in stream
+	// cleanup that causes stream to be removed too aggressively, the connection
+	// will be closd and the stream will be broken.
+	go ss.s.GracefulStop()
```

Comment:
Shouldn't we also make sure that GracefulStop() returns within a reasonable time? Because if it doesn't return by the end of the test, I'm wondering if the test would fail citing "hanging goroutines".

Reply:
Done.

Reply:
LGTM